nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/InferAvroSchema.java

```diff
@@ -0,0 +1,454 @@
+                    break;
+                default:
+                    //Shouldn't be possible but just in case
+                    session.transfer(original, REL_UNSUPPORTED_CONTENT);
+                    break;
+            }
+
+
+            if (StringUtils.isNotEmpty(avroSchema.get())) {
+
+                String destination = context.getProperty(SCHEMA_DESTINATION).getValue();
+                FlowFile avroSchemaFF = null;
+
+                switch (destination) {
+                    case DESTINATION_ATTRIBUTE:
+                        avroSchemaFF = session.putAttribute(session.clone(original), AVRO_SCHEMA_ATTRIBUTE_NAME, avroSchema.get());
+                        //Leaves the original CoreAttributes.MIME_TYPE in place.
+                        break;
+                    case DESTINATION_CONTENT:
+                        avroSchemaFF = session.write(session.create(), new OutputStreamCallback() {
+                            @Override
+                            public void process(OutputStream out) throws IOException {
+                                out.write(avroSchema.get().getBytes());
+                            }
+                        });
+                        avroSchemaFF = session.putAttribute(avroSchemaFF, CoreAttributes.MIME_TYPE.key(), AVRO_MIME_TYPE);
+                        break;
+                    default:
+                        break;
+                }
+
+                //Transfer the sessions.
+                avroSchemaFF = session.putAttribute(avroSchemaFF, CoreAttributes.FILENAME.key(), (original.getAttribute(CoreAttributes.FILENAME.key()) + AVRO_FILE_EXTENSION));
+                session.transfer(avroSchemaFF, REL_SUCCESS);
+            } else {
+                //If the avroSchema is null then the content type is unknown and therefore unsupported
+                session.transfer(session.clone(original), REL_UNSUPPORTED_CONTENT);
+            }
+
+            session.transfer(original, REL_ORIGINAL);
+
+        } catch (Exception ex) {
+            getLogger().error("Failed to infer Avro schema! {}", ex);
+            session.transfer(original, REL_FAILURE);
+        }
+    }
+
+
+    /**
+     * Infers the Avro schema from the input Flowfile content. To infer an Avro schema for CSV content a header line is
+     * required. You can configure the processor to pull that header line from the first line of the CSV data if it is
+     * present OR you can manually supply the desired header line as a property value.
+     *
+     * @param inputFlowFile
+     *  The original input FlowFile containing the CSV content as it entered this processor.
+     *
+     * @param context
+     *  ProcessContext to pull processor configurations.
+     *
+     * @param session
+     *  ProcessSession to transfer FlowFiles
+     */
+    private AtomicReference<String> inferAvroSchemaFromCSV(final FlowFile inputFlowFile, final ProcessContext context, final ProcessSession session) {
```

Comment:
Why are we returning AtomicReference<String> instead of just String here? It appears that when we get the value above, we immediately just call get() on it